enhancements/machine-api/control-plane-set.md

```diff
@@ -0,0 +1,585 @@
+- Should we teach the etcd operator to remove the protection mechanism when the ControlPlaneSet is configured with this
+  strategy?
+- To minimise risk, should we allow this strategy only on certain platforms? (Eg disallow the strategy on public clouds)
+
+##### The OnDelete update strategy
+
+The OnDelete strategy mirrors the OnDelete strategy familiar to users from StatefulSets.
+When a change is required, any logic for rolling out the Machine will be paused until the ControlPlaneSet controller
+observes a deleted Machine.
+When a Machine is marked for deletion, it will create a new Machine based on the current spec.
+It will then proceed with the replacement as normal.
+
+This strategy will allow for explicit control for end users to decide when their Control Plane Machines are updated.
+In particular it will also allow them to have different specs across their control plane.
+While in the normal case this is discouraged, this could be used for short periods to test new configuration before
+rolling the updated configuration to all Control Plane Machines.
+
+#### Protecting etcd quorum during update operations
+
+With the introduction of the [etcd protection enhancement](https://github.com/openshift/enhancements/pull/943), the
+ControlPlaneSet does not need to observe anything etcd related during scaling operations.
+In particular, it is the etcd operators responsibility to add Machine deletion hooks to prevent Control Plane Machines
+from being removed until they are no longer needed.
+When the ControlPlaneSet controller observes that a newly created Machine is ready, it will delete the old Control
+Plane Machine signalling to the etcd operator to switch the membership of the etcd cluster between the old and new
+instances.
+Once the membership has been switched, the Machine deletion hook will be removed on the old Machine, allowing it to be
+removed by the Machine controller in the normal way.
+
+#### Removing/disabling the ControlPlaneSet
+
+As some users may want to remove or disable the ControlPlaneSet, and we have no bulletproof way to prevent a user
+deleting the ControlPlaneSet, a finalizer will be placed on the ControlPlaneSet to allow the controller to ensure a
+safe removal of the ControlPlaneSet, while leaving the Control Plane Machines in place.
+
+Notably it will need to ensure that there are no owner references on the Machines pointing to the ControlPlaneSet
+instances. This will prevent the garbage collector from removing the Control Plane Machines when the ControlPlaneSet is
+deleted.
+
+If users later wish to reinstall the ControlPlaneSet, they are free to do so.
+
+#### Installing a ControlPlaneSet within an existing cluster
+
+When adding a ControlPlaneSet to a new cluster, the end user will need to define the ControlPlaneSet resource by
+copying the existing Control Plane Machine ProviderSpecs.
+Once this is copied, they should remove the failure domain and add the desired failure domains to the `FailureDomains`
```

Comment:
> Is there a possibility adoption could be part of the ControlPlaneSet creation, triggered by some specific condition such as an annotation?

Could you expand on your concern here, I'm not sure exactly what you're saying. The idea is that the adoption process happens when the CPS is created, so I would expect SRE to apply a new CPS to each individual cluster (possibly in batches) if they want to take advantage of the CPS on older clusters. My recommendation though is that we focus on new clusters where this functionality will be included from the start

> Having this as a part of the CPS operator would allow the deletion and recreation of the CR (previous section) to be nearly trivial from an administrator perspective.

So your suggestion is that if you create an empty CPS, the ControlPlaneSet operator should populate the spec for you? This feels like an anti-pattern for Kube and the idea of declarative APIs. I'm hesitant about us suggesting this as there's no guarantee that the existing configs are accurate, eg if the machines were on a UPI cluster (they should have been removed but maybe there were), at that point, we have introduced a spec that is invalid. If we infer it and get it wrong, that's on us, if a user specs it and gets it wrong, that's on them 🤔 

We should also bear in mind the scope of this project, adding adoption into the mix and having to deal with the safety aspects of that will significantly expand the scope and extend the development time which is likely to already be tight if we want to get something running for this within 4.11

Reply:
Manually managing these for the managed OCP fleet isn't an option.  But understood the concern with automatically creating the spec for CPS.  For my use case though it's IPI 100% of the time.  I am also not suggesting we populate CPS all the time, would it be a more tenable solution if it were an opt-in?

Reply:
I was expecting that the managed fleet would come up with some automation for creating the CPS, I don't think this warrants being in the core of the product though and I'm concerned about the scope creep.

We could possible have an opt-in adoption mechanism, but if we were to do that, that would have to be a stretch goal and not part of the MVP IMO, I wouldn't want to jeopardise the main project for this. I'm also not sure I'd want to expose this to all users which makes me think again, this should be something separate, what is your hesitation about running some job on the cluster to create the CPS?

Reply:
I can't speak for the use case across all OCP customers but it seems like a useful pattern to adopt via code in the operator instead of documentation and steps executed by a customer.  There are a lot of clusters out in the wild, I presume management of masters is a useful capability.  I know from the managed OCP perspective it's been a gap that is getting more attention lately.

So I'm not necessarily reluctant to run a job to create CPS resources, though the mechanics on how we can roll that out will make it painful.  I'm asking if that's the right way to do it for OCP and our customers.

Reply:
> for example, the user could issue an oc command, and it would return a manifest they could save and then inspect before applying. that might provide a nice middle ground where it could be automated but doesn't need to be.

We talked about this during the cluster lifecycle arch call last week. There were others with concerns about having this as an automated thing in-cluster. We automate it and break the cluster, that's on us, the customer breaks it, that's on them. 

There was some support for a CLI kind of thing/script that would spit out a CPS manifest for the customer to apply if they wanted, with the caveat that they should verify it before they apply it. Though my initial stance on this was that this should be a stretch goal, certainly for the purpose of this enhancement and feature we have been focusing on the experience for new clusters, older clusters can adopt later IMO